Migrate to Gradle build system (opinionated)

janheinrichmerker opened the pull request:
This PR migrates the build from Maven to Gradle.
Besides better readable build config, this makes the build run on any machine, using the bundled, yet lightweight, Gradle wrapper.
You can read more about the benefits of Gradle here: https://gradle.org/maven-vs-gradle/

janheinrichmerker committed:
Prepare for Gradle build

janheinrichmerker committed:
Update .gigignore

janheinrichmerker committed:
Remove dependency exclusions
Gradle should be able to resolve conflicts

janheinrichmerker committed:
Update IntelliJ config

janheinrichmerker committed:
Change line breaks

janheinrichmerker committed:
Update repositories

heinrichreimer commented:
After resolving some missing dependencies (repositories are [not transitive in Gradle](https://docs.gradle.org/current/userguide/declaring_repositories.html#strict_limitation_to_declared_repositories)), it now builds, as you can see here: https://jitpack.io/#heinrichreimer/trec-near-duplicates/d9aae6c2f9

mam10eks commented:
I am open to change the build system to gradle, but I will need some time to review this merge request since the final, fat jar must be able to run in our spark-cluster. (and also some scripts need to be changed, e.g. the Makefile and minor changes in the scripts in src/main/bash)

janheinrichmerker committed:
Include dependencies from non-standard repos in JAR for publishing

janheinrichmerker committed:
Include dependencies from non-standard repos in JAR for publishing
The JAR now includes all dependencies, that are not available either in jCenter or Maven Central repositories.

heinrichreimer commented:
Ah, I see. Didn't know about any other build prerequisites.

heinrichreimer commented:
Now all dependencies from third-party/local repos are included in the release JAR.

janheinrichmerker committed:
Update Makefile to use Gradle commands

janheinrichmerker commented on Makefile:
```diff
@@ -145,10 +145,3 @@ clean-stuff:
 	rm -Rf /tmp/qrels* &&\
 	rm -Rf /tmp/tmp-unzipped* &&\
 	rm -Rf /tmp/.run*
-
-install-third-party: checkout-submodules
-	mvn -f third-party/jtreceval/ clean install
-
-checkout-submodules:
-	git submodule update --init --recursive
-
```
Those are not needed anymore, as there's no submodule to checkout and build.

janheinrichmerker replied:
See #1 😃

janheinrichmerker commented on Makefile:
```diff
@@ -129,7 +129,7 @@ evaluate-clueweb12: install
 
 install: install-third-party
-	mvn clean install
+	./gradlew clean publishToMavenLocal
```
The `publishToMavenLocal` command builds the Gradle project and publishes the JAR artifact to the local Maven repo, effectively what `mvn install` did.

janheinrichmerker committed:
Add missing dependencies

janheinrichmerker committed:
Update dependency configurations

janheinrichmerker commented on build.gradle.kts:
```diff
@@ -0,0 +1,82 @@
+    implementation("net.jodah:failsafe:2.0.1")
+    implementation("org.apache.solr:solr-core:8.1.1")
+    implementation("com.github.mam10eks:jtreceval:-SNAPSHOT")
+    implementation("org.apache.commons:commons-lang3:3.9")
```
See #4

heinrichreimer commented:
Closing due to inactivity
